File: bindings/ruby/test/test_chunk.rb
# frozen_string_literal: true

require_relative "helper"

# Token-window chunking tests for the Ruby binding. Mirrors
# bindings/python/tests/test_chunk.py. Run over a byte_id pipeline (each
# input byte = one token) so chunk boundaries are predictable:
# "abcdefghij" is 10 tokens, one per byte.
class TestChunk < Minitest::Test
  include ZtokTestFixtures

  def with_byte_pipeline
    pipe = Ztok::Pipeline.byte_id
    begin
      yield pipe
    ensure
      pipe.close
    end
  end

  def test_chunk_non_overlapping
    with_byte_pipeline do |pipe|
      chunks = pipe.chunk("abcdefghij", max_tokens: 4, overlap: 0)
      # 10 tokens / window 4, stride 4 -> [0,4) [4,8) [8,10).
      assert_equal 3, chunks.length
      # [token_start, token_end, byte_start, byte_end, n_ids]
      want = [[0, 4, 0, 4, 4], [4, 8, 4, 8, 4], [8, 10, 8, 10, 2]]
      chunks.zip(want).each do |c, (ts, te, bs, be, n)|
        assert_equal [ts, te], [c.token_start, c.token_end]
        assert_equal [bs, be], [c.byte_start, c.byte_end]
        assert_equal n, c.ids.length
      end
    end
  end

  def test_chunk_overlap
    with_byte_pipeline do |pipe|
      chunks = pipe.chunk("abcdefghij", max_tokens: 4, overlap: 2)
      assert_operator chunks.length, :>=, 2
      # stride = 2, so the last 2 ids of chunk[i] equal the first 2 of
      # chunk[i+1].
      chunks.each_cons(2) do |a, b|
        if a.ids.length >= 2 && b.ids.length >= 2
          assert_equal a.ids[-2, 2], b.ids[0, 2]
        end
      end
    end
  end

  def test_chunk_empty_and_bad_args
    with_byte_pipeline do |pipe|
      assert_equal [], pipe.chunk("", max_tokens: 4)
      assert_raises(Ztok::InvalidInputError) do
        pipe.chunk("abc", max_tokens: 0)
      end
      assert_raises(Ztok::InvalidInputError) do
        pipe.chunk("abc", max_tokens: 4, overlap: 4)
      end
    end
  end
end

File: bindings/python/tests/test_chunk.py
"""Token-window chunking tests for the Python binding.

Run over a byte_id pipeline (each input byte = one token) so chunk
boundaries are predictable: "abcdefghij" is 10 tokens, one per byte.
"""

from __future__ import annotations

import pytest

import ztok


@pytest.fixture
def byte_pipeline() -> ztok.Pipeline:
    pipe = ztok.Pipeline.byte_id()
    yield pipe
    pipe.close()


def test_chunk_non_overlapping(byte_pipeline: ztok.Pipeline) -> None:
    chunks = byte_pipeline.chunk("abcdefghij", max_tokens=4, overlap=0)
    # 10 tokens / window 4, stride 4 -> [0,4) [4,8) [8,10).
    assert len(chunks) == 3
    want = [(0, 4, 0, 4, 4), (4, 8, 4, 8, 4), (8, 10, 8, 10, 2)]
    for c, (ts, te, bs, be, n) in zip(chunks, want):
        assert (c.token_start, c.token_end) == (ts, te)
        assert (c.byte_start, c.byte_end) == (bs, be)
        assert len(c.ids) == n


def test_chunk_overlap(byte_pipeline: ztok.Pipeline) -> None:
    chunks = byte_pipeline.chunk("abcdefghij", max_tokens=4, overlap=2)
    assert len(chunks) >= 2
    # stride = 2, so the last 2 ids of chunk[i] equal the first 2 of
    # chunk[i+1].
    for a, b in zip(chunks, chunks[1:]):
        if len(a.ids) >= 2 and len(b.ids) >= 2:
            assert a.ids[-2:] == b.ids[:2]


def test_chunk_empty_and_bad_args(byte_pipeline: ztok.Pipeline) -> None:
    assert byte_pipeline.chunk("", max_tokens=4) == []
    with pytest.raises(ztok.ZtokInvalidInputError):
        byte_pipeline.chunk("abc", max_tokens=0)
    with pytest.raises(ztok.ZtokInvalidInputError):
        byte_pipeline.chunk("abc", max_tokens=4, overlap=4)
